third_party/move/move-compiler-v2/src/pipeline/variable_coalescing.rs

```diff
@@ -0,0 +1,239 @@
+            LiveIntervalEvent::End(_, offset) => *offset,
+        }
+    }
+}
+
+pub struct VariableCoalescing {}
+
+impl VariableCoalescing {
+    /// Compute the live intervals of locals in the given function target.
+    /// The result is a vector of live intervals, where the index of the vector is the local.
+    /// If a local has `None` as its live interval, we can ignore the local for the coalescing
+    /// transformation (eg., because it is borrowed or because it is never used): it implies
+    /// that it is the trivial live interval.
+    fn live_intervals(target: &FunctionTarget) -> Vec<Option<LiveInterval>> {
+        let LiveVarAnnotation(live_var_infos) = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("live var annotation is a prerequisite");
+        // Note: we currently exclude all the variables that are borrowed from participating in this
+        // transformation, which is safe. However, we could be more precise in this regard.
+        let borrowed_locals = target.get_borrowed_locals();
+        // Initially, all locals have trivial live intervals.
+        let mut live_intervals = std::iter::repeat_with(|| None)
+            .take(target.get_local_count())
+            .collect::<Vec<_>>();
+        for (offset, live_var_info) in live_var_infos.iter() {
+            live_var_info
+                .after
+                .keys()
+                .chain(live_var_info.before.keys())
+                .filter(|local| !borrowed_locals.contains(local))
+                .for_each(|local| {
+                    // non-borrowed local that is live before and/or after the code offset.
+                    let interval =
+                        live_intervals[*local].get_or_insert_with(|| LiveInterval::new(*offset));
+                    interval.include(*offset);
+                });
+        }
+        live_intervals
+    }
+
+    /// Compute the sorted live interval events of locals in the given function target.
+    /// See implementation comments for the sorting order.
+    fn sorted_live_interval_events(target: &FunctionTarget) -> Vec<LiveIntervalEvent> {
+        let live_intervals = Self::live_intervals(target);
+        let mut live_interval_events = vec![];
+        for (local, interval) in live_intervals.into_iter().enumerate() {
+            if let Some(interval) = interval {
+                live_interval_events.push(LiveIntervalEvent::Begin(
+                    local as TempIndex,
+                    interval.begin,
+                    (interval.end - interval.begin) as usize,
+                ));
+                live_interval_events.push(LiveIntervalEvent::End(local as TempIndex, interval.end));
+            }
+        }
+        live_interval_events.sort_by(|a, b| {
+            use LiveIntervalEvent::*;
+            match (a, b) {
+                // Sort events based on their code offsets (lower offset comes first).
```

Comment:
This code block feels like it would be easier to read if you pulled all the comments out into a separate block.  The code isn't that hard to understand once you get the gist of it, except that it's buried in comments.

Reply:
This is one place where `Range<CodeOffset>` may have helped?

Reply:
done (pulling comments out to the beginning of the function, instead of interspersing with code).

`Range` does not help here as we are dealing with discrete events (each with one extreme of a range). However, I did end up using `RangeInclusive` in the `LiveInterval`.